rclrs/src/node/client.rs

```diff
@@ -0,0 +1,268 @@
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub fn async_send_request_with_callback<'a, M: MessageCow<'a, T::Request>, F>(
+        &self,
+        message: M,
+        callback: F,
+    ) -> Result<(), RclrsError>
+    where
+        F: FnOnce(&T::Response) + 'static + Send,
+    {
+        let rmw_message = T::Request::into_rmw_message(message.into_cow());
+        let mut sequence_number = -1;
+        let ret = unsafe {
+            // SAFETY: The request type is guaranteed to match the client type by the type system.
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        };
+        let requests = &mut *self.requests.lock();
+        requests.insert(sequence_number, Box::new(callback));
+        ret.ok()
+    }
+
+    /// Send a request with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub async fn call_async<'a, R: MessageCow<'a, T::Request>>(
+        &self,
+        request: R,
+    ) -> Result<T::Response, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+    {
+        let rmw_message = T::Request::into_rmw_message(request.into_cow());
+        let mut sequence_number = -1;
+        unsafe {
+            // SAFETY: The request type is guaranteed to match the client type by the type system.
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        }
+        .ok()?;
+        let (tx, rx) = oneshot::channel::<T::Response>();
+        self.futures.lock().insert(sequence_number, tx);
+        Ok(rx.await.unwrap())
+    }
+
+    /// Ask RMW for the data
+    ///
+    /// +----------------------+
+    /// | rclrs::take_response |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v-----------+
+    /// |   rcl_take_response  |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v----------+
+    /// |      rmw_take       |
+    /// +---------------------+
```

Comment:
This diagram is nice for rclrs devs, but not that helpful for end users. I propose to put it in regular comments, not doc comments, like `Subscription::take()`.

Reply:
`Subscription::take()` does have the similar diagram, not a regular comment.

Reply:
What I mean is it uses double slashes, not triple slashes.

Reply:
Same for the service.

Reply:
Done in 4191195